content-src/lib/selectLayoutRender.js

```diff
@@ -1,79 +1,83 @@
-import {createSelector} from "reselect";
+export const selectLayoutRender = (state, rickRollCache) => {
+  const {layout, feeds, spocs} = state;
+  let spocIndex = 0;
+  let bufferRollCache = [];
 
-export const selectLayoutRender = createSelector(
-  // Selects layout, feeds, spocs so that we only recompute if
-  // any of these values change.
-  [
-    state => state.DiscoveryStream.layout,
-    state => state.DiscoveryStream.feeds,
-    state => state.DiscoveryStream.spocs,
-  ],
+  // rickRollCache stores random probability values for each spoc position. This cache is empty
+  // on page refresh and gets filled with random values on first render inside maybeInjectSpocs.
+  const isFirstRun = !rickRollCache.length;
 
-  // Adds data to each component from feeds. This function only re-runs if one of the inputs change.
-  // TODO: calculate spocs
-  function layoutRender(layout, feeds, spocs) {
-    let spocIndex = 0;
-
-    function maybeInjectSpocs(data, spocsConfig) {
-      if (data &&
-          spocsConfig && spocsConfig.positions && spocsConfig.positions.length &&
-          spocs.data.spocs && spocs.data.spocs.length) {
-        const recommendations = [...data.recommendations];
-        for (let position of spocsConfig.positions) {
-          let rickRoll = Math.random();
-          if (spocs.data.spocs[spocIndex] && rickRoll <= spocsConfig.probability) {
-            recommendations.splice(position.index, 0, spocs.data.spocs[spocIndex++]);
+  function maybeInjectSpocs(data, spocsConfig) {
+    if (data &&
+        spocsConfig && spocsConfig.positions && spocsConfig.positions.length &&
+        spocs.data.spocs && spocs.data.spocs.length) {
+      const recommendations = [...data.recommendations];
+      for (let position of spocsConfig.positions) {
+        // Cache random number for a position
+        let rickRoll;
+        if (isFirstRun) {
+          rickRoll = Math.random();
+          rickRollCache.push(rickRoll);
+        } else {
+          rickRoll = rickRollCache.shift();
+          bufferRollCache.push(rickRoll);
+          if (!rickRollCache.length) {
+            rickRollCache.push(...bufferRollCache);
```

Comment:
Possible to do the above check and push right after the `maybeInjectSpocs` call on line 66?